### PR TITLE
initial fastify instrumentation

### PR DESCRIPTION
**Pull request by toshok**
Introduce `fastify` auto-instrumentation.  It registers two hooks before anything in user-code is run, one `onRequest` to start the trace, and one `onResponse` to send it.  Given just how pervasively async the api is and how things are registered, it's not as easy to ensure that context gets propagated correctly through hooks, so instead we store off the trace context in a Map and fetch it for every hook.  This means that there may be blips where context is lost, but each hook/handler should begin anew with the context.

Things this PR probably needs to include:

- [x] hooks
- [x] route handlers (`.get`, `.put`, etc)
- [ ] middleware - skipping this for now.  if we have requests, can add it via another PR.

### CODE REVIEW

File: lib/instrumentation/fastify.js [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+      finishersByRequest.set(request, boundFinisher);
+      trackedByRequest.set(request, tracker.getTracked());
+
+      next();
+    });
+
+    app.addHook("onResponse", (request, reply, next) => {
+      // calculate total time for the request and send the root span
+      debug("in onResponse");
```
Comment (dsoo) [inline lib/instrumentation/fastify.js]: Do we want to leave the various debug spam in?

File: lib/instrumentation/fastify.js [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+      const finisher = finishersByRequest.get(request);
+      if (finisher) {
+        debug("finishing");
+        finisher(reply);
+      }
+
+      next();
+    });
+
+    // now that we've added our special hooks, instrument the addHook method such that
```
Comment (dsoo) [inline lib/instrumentation/fastify.js]: Incomplete comment?

File: lib/instrumentation/fastify.js [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+          let request = args[0];
+
+          let tracked = trackedByRequest.get(request);
+          tracker.setTracked(tracked);
+
+          let span = api.startSpan({
+            name: `${hookName} hook`,
+            [schema.EVENT_TYPE]: "fastify",
+            [schema.PACKAGE_VERSION]: opts.packageVersion,
+            // XXX toshok more here
```
Comment (dsoo) [inline lib/instrumentation/fastify.js]: ???

File: lib/instrumentation/fastify.js [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+        let wrappedHandler = function(request, reply) {
+          let tracked = trackedByRequest.get(request);
+          tracker.setTracked(tracked);
+
+          let span = api.startSpan({
+            name: `handler: ${method} ${handlerUrl}`,
+            [schema.EVENT_TYPE]: "fastify",
+            [schema.PACKAGE_VERSION]: opts.packageVersion,
+            // XXX toshok more here
```
Comment (dsoo) [inline lib/instrumentation/fastify.js]: ???

File: lib/instrumentation/fastify.js (hunk 1 above)
Comment (toshok) [reply lib/instrumentation/fastify.js]: ah, definitely not. 👍

**Comment by wolfeidau** [comment]
Is there any examples of how to use this?

I can't find it in the documentation.

**Pull request merged**